library/core/src/slice/cmp.rs

```diff
@@ -60,7 +60,22 @@ where
             return false;
         }
 
-        self.iter().zip(other.iter()).all(|(x, y)| x == y)
+        let (chunks_a, residual_a) = self.as_chunks::<4>();
+        let (chunks_b, residual_b) = other.as_chunks::<4>();
+
+        // check the residual first to bail out fast if there's a mismatch and comparisons
+        // happen to be expensive
+        let mut result = residual_a.into_iter().zip(residual_b).all(|(a, b)| a == b);
+
+        // iter.all short-circuits which means the backend can't unroll the loop due to early exits.
+        // So we unroll it manually.
+        result = result
+            && chunks_a
+                .iter()
+                .zip(chunks_b.iter())
+                .all(|(a, b)| (a[0] == b[0]) & (a[1] == b[1]) & (a[2] == b[2]) & (a[3] == b[3]));
```

Comment:
I guess this is `&` because of things like https://github.com/rust-lang/rust/issues/105259#issuecomment-1442496307 ?

It seems wrong to do this for general arbitrarily-expensive `==`.  Is there a way it could be written to pre-load the values so that LLVM knows it's *allowed* to remove the short-circuit, but it's still up to the optimizer to decide if it's worth doing?

Reply:
Yeah, there are a bunch of other issues about `&&` not doing the right thing too. I initially used `flag &= ...` but that unrolled to a sequence of cmp + jump pairs. Which was faster too but a lot more instructions.

They're all references anyway, so we can't really "load" them. We could special-case it for `T: Copy` but that wouldn't make a difference for `&str` or similar.

I'm considering to peel the first chunk (in addition to the residual) and do it the conventional way if perf results are bad. This should hopefully catch a lot of non-equal results and mostly do the unrolled variant

Reply:
I tried preloading all the values with `let _ = ManuallyDrop::new(ptr::read(self.get_unchecked(i)));` etc. and then using `&&` but it didn't help.